### PR TITLE
Add more fairness metrics

### PR DESCRIPTION
**Pull request by cuiyuebing**
Add more fairness metrics

### CODE REVIEW

File: federatedscope/core/monitors/monitor.py [hunk 1]
```diff
@@ -376,6 +376,24 @@ def format_eval_res(self,
                                 all_res.size // 10]
                             new_results[f"{key}_top_decile"] = all_res[
                                 all_res.size * 9 // 10]
+                            # log more fairness metrics
+                            
+                            # min and max
+                            new_results[f"{key}_min"] = np.mean(all_res[0])
```
Comment (yxdyc) [inline federatedscope/core/monitors/monitor.py:382]: why np.mean()?   all_res[0] or all_res[-1] should be only one element, i.e., the performance of one client

File: federatedscope/core/monitors/monitor.py [hunk 2]
```diff
@@ -376,6 +376,24 @@ def format_eval_res(self,
...
+                            new_results[f"{key}_max"] = np.mean(all_res[-1])
+
+                            # best and worst 10%
+                            new_results[f"{key}_worst10%"] = np.mean(all_res[:
```
Comment (yxdyc) [inline federatedscope/core/monitors/monitor.py:386]: similar to the above question.  if there is no other special consideration, the best/worst10% will be the same as the metric top/bottom_decile that have been calculated in line 375/377

Comment (yxdyc) [reply federatedscope/core/monitors/monitor.py:386]: Besides, the term "worst"/"best" is conditioned on the metric. We may need to maintain a flag such as `the_larger_the_better`

File: federatedscope/core/monitors/monitor.py [hunk 3]
```diff
@@ -376,6 +376,24 @@ def format_eval_res(self,
...
+                                                                             all_res.size // 10])
+                            new_results[f"{key}_best10%"] = np.mean(all_res[
+                                                                    all_res.size * 9 // 10:])
+
+                            # cosine similarity between the performance distribution and 1
+                            new_results[f"{key}_cos1"] = np.mean(all_res) / (np.sqrt(np.mean(all_res ** 2)))
+
+                            # entropy of performance distribution
+                            new_results[f"{key}_entropy"] = np.sum(
+                                -all_res / np.sum(all_res + 1e-9) * (np.log((all_res + 1e-9) / np.sum(all_res + 1e-9))))
```
Comment (yxdyc) [inline federatedscope/core/monitors/monitor.py:396]: perhaps
```
all_res = (all_res + 1e-9) / np.sum(all_res + 1e-9)
new_results[f"{key}_entropy"] = np.sum(-all_res * np.log(all_res))
```

**Comment by yxdyc** [comment]
LGTM

**Review by yxdyc — APPROVED** [review]

File: federatedscope/core/monitors/monitor.py [hunk 4]
```diff
@@ -376,6 +376,20 @@ def format_eval_res(self,
...
+                            #bottom and top 10%
+                            new_results[f"{key}_bottom10%"]=np.mean(all_res[:
+                                all_res.size  // 10])
+                            new_results[f"{key}_top10%"]=np.mean(all_res[
+                                all_res.size * 9 // 10:])
+                            #cosine similarity between the performance distribution and 1
+                            new_results[f"{key}_cos1"]=np.mean(all_res)/(np.sqrt(np.mean(all_res**2)))
+                            #entropy of performance distribution
+                            all_res_preprocessed=all_res+1e-9
+                            new_results[f"{key}_entropy"]=np.sum(-all_res_preprocessed/np.sum(all_res_preprocessed)*(np.log((all_res_preprocessed)/np.sum(all_res_preprocessed))))
```
Comment (Osier-Yi) [inline federatedscope/core/monitors/monitor.py:392]: Please update the comments in line 318 according to the new metrics added: 
            'Results_fairness': {
                'test_correct': 3356,      'test_total': 5000,
                'test_avg_loss_std': 0.04, 'test_avg_loss_bottom_decile':
                0.52, 'test_avg_loss_top_decile': 0.64,
                'test_acc_std': 0.06,      'test_acc_bottom_decile': 0.60,
                'test_acc_top_decile': 0.75,
                'test_loss_std': 214.17,   'test_loss_bottom_decile':
                2644.64, 'test_loss_top_decile': 3241.23
                },
            }

**Comment by Osier-Yi** [comment]
LGTM

**Review by Osier-Yi — APPROVED** [review]

**Pull request merged**